File: lib/types/property.ts
/**
 * Property Types
 * TypeScript types and interfaces for the properties table
 */

// Property type enum
export type PropertyType =
  | 'apartment'
  | 'house'
  | 'studio'
  | 'coliving'
  | 'shared_room'
  | 'private_room'
  | 'entire_place';

// Property status enum
export type PropertyStatus =
  | 'draft'
  | 'published'
  | 'archived'
  | 'rented'
  | 'under_review';

// Amenity types (common amenities)
export type PropertyAmenity =
  | 'wifi'
  | 'parking'
  | 'elevator'
  | 'balcony'
  | 'garden'
  | 'gym'
  | 'laundry'
  | 'dishwasher'
  | 'washing_machine'
  | 'dryer'
  | 'air_conditioning'
  | 'heating'
  | 'kitchen'
  | 'furnished'
  | 'pets_allowed'
  | 'smoking_allowed'
  | 'wheelchair_accessible'
  | 'security'
  | 'concierge'
  | 'pool'
  | 'terrace';

// Main Property interface
export interface Property {
  id: string;
  owner_id: string;

  // Basic Information
  title: string;
  description: string | null;
  property_type: PropertyType;

  // Location
  address: string;
  city: string;
  neighborhood: string | null;
  postal_code: string;
  country: string;
  latitude: number | null;
  longitude: number | null;

  // Property Details
  bedrooms: number;
  bathrooms: number;
  total_rooms: number | null;
  surface_area: number | null; // in square meters
  floor_number: number | null;
  total_floors: number | null;
  furnished: boolean;

  // Pricing
  monthly_rent: number;
  charges: number;
  deposit: number | null;

  // Availability
  available_from: string | null; // ISO date string
  available_until: string | null; // ISO date string
  minimum_stay_months: number;
  maximum_stay_months: number | null;
  is_available: boolean;

  // Amenities
  amenities: PropertyAmenity[];

  // Rules & Preferences
  smoking_allowed: boolean;
  pets_allowed: boolean;
  couples_allowed: boolean;
  children_allowed: boolean;

  // Images
  images: string[]; // Array of image URLs
  main_image: string | null;

  // Status & Metadata
  status: PropertyStatus;
  views_count: number;
  applications_count: number;
  favorites_count: number;

  // Timestamps
  created_at: string;
  updated_at: string;
  published_at: string | null;
  archived_at: string | null;
}

// Property creation input (for forms)
export interface PropertyCreateInput {
  title: string;
  description?: string;
  property_type: PropertyType;

  // Location
  address: string;
  city: string;
  neighborhood?: string;
  postal_code: string;
  country?: string;
  latitude?: number;
  longitude?: number;

  // Property Details
  bedrooms: number;
  bathrooms: number;
  total_rooms?: number;
  surface_area?: number;
  floor_number?: number;
  total_floors?: number;
  furnished?: boolean;

  // Pricing
  monthly_rent: number;
  charges?: number;
  deposit?: number;

  // Availability
  available_from?: string;
  available_until?: string;
  minimum_stay_months?: number;
  maximum_stay_months?: number;

  // Amenities
  amenities?: PropertyAmenity[];

  // Rules
  smoking_allowed?: boolean;
  pets_allowed?: boolean;
  couples_allowed?: boolean;
  children_allowed?: boolean;

  // Images
  images?: string[];
  main_image?: string;
}

// Property update input (partial)
export type PropertyUpdateInput = Partial<PropertyCreateInput> & {
  id: string;
};

// Property filters for search
export interface PropertyFilters {
  city?: string;
  neighborhood?: string;
  property_type?: PropertyType;
  min_price?: number;
  max_price?: number;
  bedrooms?: number;
  bathrooms?: number;
  amenities?: PropertyAmenity[];
  is_furnished?: boolean;
  pets_allowed?: boolean;
  smoking_allowed?: boolean;
  available_from?: string;
  minimum_stay_months?: number;
  maximum_stay_months?: number;
}

// Property with owner info (for listings)
export interface PropertyWithOwner extends Property {
  owner: {
    id: string;
    email: string;
    full_name?: string;
    avatar_url?: string;
    phone?: string;
  };
}

// Property stats (for owner dashboard)
export interface PropertyStats {
  total_properties: number;
  published_properties: number;
  draft_properties: number;
  rented_properties: number;
  total_views: number;
  total_applications: number;
  total_favorites: number;
  avg_applications_per_property: number;
}

// Property card (minimal info for lists)
export interface PropertyCard {
  id: string;
  title: string;
  property_type: PropertyType;
  city: string;
  neighborhood: string | null;
  monthly_rent: number;
  bedrooms: number;
  bathrooms: number;
  surface_area: number | null;
  main_image: string | null;
  images: string[];
  status: PropertyStatus;
  views_count: number;
  favorites_count: number;
  is_available: boolean;
  published_at: string | null;
}

// Property validation constraints
export const PROPERTY_CONSTRAINTS = {
  title: {
    minLength: 10,
    maxLength: 255,
  },
  description: {
    minLength: 50,
    maxLength: 5000,
  },
  bedrooms: {
    min: 0,
    max: 20,
  },
  bathrooms: {
    min: 0,
    max: 10,
  },
  total_rooms: {
    min: 0,
    max: 50,
  },
  surface_area: {
    min: 1,
    max: 10000,
  },
  monthly_rent: {
    min: 0,
    max: 1000000,
  },
  charges: {
    min: 0,
    max: 10000,
  },
  deposit: {
    min: 0,
    max: 50000,
  },
  minimum_stay_months: {
    min: 1,
    max: 24,
  },
  maximum_stay_months: {
    min: 1,
    max: 60,
  },
  images: {
    maxCount: 20,
    maxSizeBytes: 10 * 1024 * 1024, // 10MB
    allowedTypes: ['image/jpeg', 'image/jpg', 'image/png', 'image/webp', 'image/gif'],
  },
} as const;

// Property type labels (for UI)
export const PROPERTY_TYPE_LABELS: Record<PropertyType, string> = {
  apartment: 'Apartment',
  house: 'House',
  studio: 'Studio',
  coliving: 'Coliving Space',
  shared_room: 'Shared Room',
  private_room: 'Private Room',
  entire_place: 'Entire Place',
};

// Property status labels (for UI)
export const PROPERTY_STATUS_LABELS: Record<PropertyStatus, string> = {
  draft: 'Draft',
  published: 'Published',
  archived: 'Archived',
  rented: 'Rented',
  under_review: 'Under Review',
};

// Amenity labels (for UI)
export const AMENITY_LABELS: Record<PropertyAmenity, string> = {
  wifi: 'Wi-Fi',
  parking: 'Parking',
  elevator: 'Elevator',
  balcony: 'Balcony',
  garden: 'Garden',
  gym: 'Gym',
  laundry: 'Laundry',
  dishwasher: 'Dishwasher',
  washing_machine: 'Washing Machine',
  dryer: 'Dryer',
  air_conditioning: 'Air Conditioning',
  heating: 'Heating',
  kitchen: 'Kitchen',
  furnished: 'Furnished',
  pets_allowed: 'Pets Allowed',
  smoking_allowed: 'Smoking Allowed',
  wheelchair_accessible: 'Wheelchair Accessible',
  security: 'Security',
  concierge: 'Concierge',
  pool: 'Pool',
  terrace: 'Terrace',
};

// Helper function to format price
export function formatPrice(price: number, currency: string = '€'): string {
  return `${currency}${price.toLocaleString('fr-FR', { minimumFractionDigits: 0, maximumFractionDigits: 2 })}`;
}

// Helper function to format property type
export function formatPropertyType(type: PropertyType): string {
  return PROPERTY_TYPE_LABELS[type] || type;
}

// Helper function to format status
export function formatPropertyStatus(status: PropertyStatus): string {
  return PROPERTY_STATUS_LABELS[status] || status;
}

// Helper function to format amenity
export function formatAmenity(amenity: PropertyAmenity): string {
  return AMENITY_LABELS[amenity] || amenity;
}

// Helper function to check if property is editable
export function isPropertyEditable(property: Property): boolean {
  return property.status === 'draft' || property.status === 'under_review';
}

// Helper function to check if property can be published
export function canPublishProperty(property: Property): boolean {
  return (
    property.status === 'draft' &&
    property.title.length >= PROPERTY_CONSTRAINTS.title.minLength &&
    property.description !== null &&
    property.description !== undefined &&
    property.description.length >= PROPERTY_CONSTRAINTS.description.minLength &&
    property.images.length > 0 &&
    property.monthly_rent > 0
  );
}

File: EasyCoiOS-Clean/EasyCo/EasyCo/Models/Property.swift
//
//  Property.swift
//  EasyCo
//
//  Complete property model matching web app structure
//

import Foundation

// MARK: - Property Model

struct Property: Identifiable, Codable {
    // IDs
    let id: UUID
    var ownerID: UUID

    // Basic Information
    var title: String
    var description: String?
    var propertyType: PropertyType

    // Location
    var address: String
    var city: String
    var neighborhood: String?
    var postalCode: String
    var country: String
    var latitude: Double?
    var longitude: Double?

    // Property Details
    var bedrooms: Int
    var bathrooms: Int
    var totalRooms: Int?
    var surfaceArea: Double? // in m²
    var floorNumber: Int?
    var totalFloors: Int?
    var furnished: Bool

    // Pricing
    var monthlyRent: Double
    var charges: Double
    var deposit: Double?

    // Availability
    var availableFrom: Date?
    var availableUntil: Date?
    var minimumStayMonths: Int
    var maximumStayMonths: Int?
    var isAvailable: Bool

    // Amenities
    var amenities: [PropertyAmenity]

    // Rules & Preferences
    var smokingAllowed: Bool
    var petsAllowed: Bool
    var couplesAllowed: Bool
    var childrenAllowed: Bool

    // Images
    var images: [String]
    var mainImage: String?

    // Status & Metadata
    var status: PropertyStatus
    var viewsCount: Int
    var applicationsCount: Int
    var favoritesCount: Int
    var rating: Double
    var reviewsCount: Int

    // Timestamps
    var createdAt: Date
    var updatedAt: Date
    var publishedAt: Date?
    var archivedAt: Date?

    // Matching (calculated, not stored in DB)
    var compatibilityScore: Int?
    var matchInsights: [String]?
    var isFavorited: Bool?

    // Computed property for compatibility
    var mainImageURL: String? {
        mainImage ?? images.first
    }

    // Residents (for property cards)
    var residents: [PropertyResident]?

    enum CodingKeys: String, CodingKey {
        case id, title, description, address, city, neighborhood, country
        case bedrooms, bathrooms, furnished, images, amenities
        case ownerID = "owner_id"
        case propertyType = "property_type"
        case postalCode = "postal_code"
        case latitude, longitude
        case totalRooms = "total_rooms"
        case surfaceArea = "surface_area"
        case floorNumber = "floor_number"
        case totalFloors = "total_floors"
        case monthlyRent = "monthly_rent"
        case charges, deposit
        case availableFrom = "available_from"
        case availableUntil = "available_until"
        case minimumStayMonths = "minimum_stay_months"
        case maximumStayMonths = "maximum_stay_months"
        case isAvailable = "is_available"
        case smokingAllowed = "smoking_allowed"
        case petsAllowed = "pets_allowed"
        case couplesAllowed = "couples_allowed"
        case childrenAllowed = "children_allowed"
        case mainImage = "main_image"
        case status
        case viewsCount = "views_count"
        case applicationsCount = "applications_count"
        case favoritesCount = "favorites_count"
        case rating
        case reviewsCount = "reviews_count"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case publishedAt = "published_at"
        case archivedAt = "archived_at"
        case compatibilityScore, matchInsights, isFavorited, residents
    }
}

// MARK: - Property Type

enum PropertyType: String, Codable, CaseIterable {
    case apartment = "apartment"
    case house = "house"
    case studio = "studio"
    case coliving = "coliving"
    case sharedRoom = "shared_room"
    case privateRoom = "private_room"
    case entirePlace = "entire_place"

    var displayName: String {
        switch self {
        case .apartment: return "Appartement"
        case .house: return "Maison"
        case .studio: return "Studio"
        case .coliving: return "Coliving"
        case .sharedRoom: return "Chambre partagée"
        case .privateRoom: return "Chambre privée"
        case .entirePlace: return "Logement entier"
        }
    }

    var icon: String {
        switch self {
        case .apartment: return "building.2"
        case .house: return "house"
        case .studio: return "bed.double"
        case .coliving: return "person.3"
        case .sharedRoom: return "person.2"
        case .privateRoom: return "door.left.hand.closed"
        case .entirePlace: return "house.fill"
        }
    }
}

// MARK: - Property Status

enum PropertyStatus: String, Codable, CaseIterable {
    case draft = "draft"
    case published = "published"
    case archived = "archived"
    case rented = "rented"
    case underReview = "under_review"

    var displayName: String {
        switch self {
        case .draft: return "Brouillon"
        case .published: return "Publié"
        case .archived: return "Archivé"
        case .rented: return "Loué"
        case .underReview: return "En révision"
        }
    }
}

// MARK: - Property Amenity

enum PropertyAmenity: String, Codable, CaseIterable {
    case wifi = "wifi"
    case parking = "parking"
    case elevator = "elevator"
    case balcony = "balcony"
    case garden = "garden"
    case gym = "gym"
    case laundry = "laundry"
    case dishwasher = "dishwasher"
    case washingMachine = "washing_machine"
    case dryer = "dryer"
    case airConditioning = "air_conditioning"
    case heating = "heating"
    case tv = "tv"
    case workspace = "workspace"
    case kitchen = "kitchen"
    case privateBathroom = "private_bathroom"
    case wheelchairAccessible = "wheelchair_accessible"
    case security = "security"
    case concierge = "concierge"
    case pool = "pool"
    case terrace = "terrace"

    var displayName: String {
        switch self {
        case .wifi: return "WiFi"
        case .parking: return "Parking"
        case .elevator: return "Ascenseur"
        case .balcony: return "Balcon"
        case .garden: return "Jardin"
        case .gym: return "Salle de sport"
        case .laundry: return "Buanderie"
        case .dishwasher: return "Lave-vaisselle"
        case .washingMachine: return "Machine à laver"
        case .dryer: return "Sèche-linge"
        case .airConditioning: return "Climatisation"
        case .heating: return "Chauffage"
        case .tv: return "TV"
        case .workspace: return "Espace de travail"
        case .kitchen: return "Cuisine"
        case .privateBathroom: return "Salle de bain privée"
        case .wheelchairAccessible: return "Accessible PMR"
        case .security: return "Sécurité"
        case .concierge: return "Concierge"
        case .pool: return "Piscine"
        case .terrace: return "Terrasse"
        }
    }

    var icon: String {
        switch self {
        case .wifi: return "wifi"
        case .parking: return "car.fill"
        case .elevator: return "arrow.up.arrow.down"
        case .balcony: return "square.grid.3x3.square"
        case .garden: return "leaf.fill"
        case .gym: return "figure.run"
        case .laundry: return "washer.fill"
        case .dishwasher: return "dishwasher.fill"
        case .washingMachine: return "washer.fill"
        case .dryer: return "dryer.fill"
        case .airConditioning: return "snowflake"
        case .heating: return "flame.fill"
        case .tv: return "tv.fill"
        case .workspace: return "desktopcomputer"
        case .kitchen: return "fork.knife"
        case .privateBathroom: return "shower.fill"
        case .wheelchairAccessible: return "figure.roll"
        case .security: return "lock.shield.fill"
        case .concierge: return "bell.fill"
        case .pool: return "figure.pool.swim"
        case .terrace: return "sun.max.fill"
        }
    }
}

// MARK: - Property Resident (for card display)

struct PropertyResident: Identifiable, Codable {
    let id: UUID
    var firstName: String?
    var profileImageURL: String?
    var initials: String {
        guard let firstName = firstName else { return "?" }
        return String(firstName.prefix(1)).uppercased()
    }
}

// MARK: - Mock Data

extension Property {
    static var mock: Property {
        mockProperties[0]
    }

    static var mockProperties: [Property] {
        [
            Property(
                id: UUID(),
                ownerID: UUID(),
                title: "Magnifique Colocation à Ixelles",
                description: "Belle colocation de 4 chambres dans une maison rénovée avec jardin. Ambiance conviviale et internationale.",
                propertyType: .coliving,
                address: "Rue de la Paix 42",
                city: "Ixelles",
                neighborhood: "Flagey",
                postalCode: "1050",
                country: "Belgique",
                latitude: 50.8269,
                longitude: 4.3676,
                bedrooms: 4,
                bathrooms: 2,
                totalRooms: 6,
                surfaceArea: 120,
                floorNumber: 2,
                totalFloors: 3,
                furnished: true,
                monthlyRent: 650,
                charges: 150,
                deposit: 1300,
                availableFrom: Date(),
                availableUntil: nil,
                minimumStayMonths: 6,
                maximumStayMonths: nil,
                isAvailable: true,
                amenities: [.wifi, .washingMachine, .dishwasher, .garden, .heating],
                smokingAllowed: false,
                petsAllowed: true,
                couplesAllowed: false,
                childrenAllowed: false,
                images: ["https://picsum.photos/400/300", "https://picsum.photos/401/300"],
                mainImage: "https://picsum.photos/400/300",
                status: .published,
                viewsCount: 234,
                applicationsCount: 12,
                favoritesCount: 45,
                rating: 4.8,
                reviewsCount: 23,
                createdAt: Date(),
                updatedAt: Date(),
                publishedAt: Date(),
                archivedAt: nil,
                compatibilityScore: 92,
                matchInsights: ["Même rythme de vie", "Préférences similaires"],
                isFavorited: false,
                residents: [
                    PropertyResident(id: UUID(), firstName: "Sophie", profileImageURL: nil),
                    PropertyResident(id: UUID(), firstName: "Marc", profileImageURL: nil),
                    PropertyResident(id: UUID(), firstName: "Julie", profileImageURL: nil)
                ]
            ),
            Property(
                id: UUID(),
                ownerID: UUID(),
                title: "Studio Moderne à Saint-Gilles",
                description: "Studio lumineux et entièrement meublé, proche de toutes commodités.",
                propertyType: .studio,
                address: "Avenue Jean Volders 15",
                city: "Saint-Gilles",
                neighborhood: "Parvis",
                postalCode: "1060",
                country: "Belgique",
                latitude: 50.8296,
                longitude: 4.3418,
                bedrooms: 1,
                bathrooms: 1,
                totalRooms: 2,
                surfaceArea: 45,
                floorNumber: 4,
                totalFloors: 5,
                furnished: true,
                monthlyRent: 850,
                charges: 100,
                deposit: 1700,
                availableFrom: Calendar.current.date(byAdding: .month, value: 1, to: Date()),
                availableUntil: nil,
                minimumStayMonths: 12,
                maximumStayMonths: nil,
                isAvailable: true,
                amenities: [.wifi, .elevator, .washingMachine, .balcony],
                smokingAllowed: false,
                petsAllowed: false,
                couplesAllowed: true,
                childrenAllowed: false,
                images: ["https://picsum.photos/402/300"],
                mainImage: "https://picsum.photos/402/300",
                status: .published,
                viewsCount: 156,
                applicationsCount: 8,
                favoritesCount: 28,
                rating: 4.5,
                reviewsCount: 15,
                createdAt: Date(),
                updatedAt: Date(),
                publishedAt: Date(),
                archivedAt: nil,
                compatibilityScore: 78,
                matchInsights: ["Budget compatible"],
                isFavorited: true,
                residents: nil
            )
        ]
    }
}
